cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -1039,7 +1039,15 @@ public void batchNextRunTimes(FullHttpRequest request,
   private List<BatchProgramSchedule> batchRunTimes(String namespace, Collection<? extends BatchProgram> programs,
                                                    boolean previous) throws Exception {
     List<ProgramId> programIds = programs.stream()
-      .map(p -> new ProgramId(namespace, p.getAppId(), p.getProgramType(), p.getProgramId()))
+      .map(p -> {
+        try {
+          return new ProgramId(namespace, p.getAppId(), getLatestAppVersion(new NamespaceId(namespace), p.getAppId()),
```

Comment:
hmmm, this is creating a loop call of fetching latest versions... Is it possible to refactor it using the actual batch API to call DB?

Reply:
try not to query the latest version in Handler, we should use ProgramReference here to reference the latest version of a program. and fetch the latest version in programLCService

Reply:
addressed the comments